lib/srv/server/installer/autodiscover.go

```diff
@@ -0,0 +1,600 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The last step is to configure the `teleport.yaml`.
+	// We could do this using the github.com/gravitational/teleport/lib/config package.
+	// However, that would cause the configuration to use the current running binary which is different from the binary that was just installed.
+	// That could cause problems if the versions are not compatible.
+	// To prevent creating an invalid configuration, the installed binary must be used.
+
+	labelEntries := make([]string, 0, len(nodeLabels))
+	for labelKey, labelValue := range nodeLabels {
+		labelEntries = append(labelEntries, labelKey+"="+labelValue)
+	}
+	sort.Strings(labelEntries)
+	nodeLabelsCommaSeperated := strings.Join(labelEntries, ",")
+
+	joinMethod, ok := imdsClientTypeToJoinMethod[imdsClient.GetType()]
+	if !ok {
+		return trace.BadParameter("Unsupported cloud provider: %v", imdsClient.GetType())
+	}
+
+	teleportNodeConfigureArgs := []string{"node", "configure", "--output=file://" + teleportYamlConfigurationPath,
+		fmt.Sprintf(`--proxy=%s`, shsprintf.EscapeDefaultContext(asi.ProxyPublicAddr)),
+		fmt.Sprintf(`--join-method=%s`, shsprintf.EscapeDefaultContext(string(joinMethod))),
+		fmt.Sprintf(`--token=%s`, shsprintf.EscapeDefaultContext(asi.TokenName)),
+		fmt.Sprintf(`--labels=%s`, shsprintf.EscapeDefaultContext(nodeLabelsCommaSeperated)),
+	}
+	if asi.AzureClientID != "" {
+		teleportNodeConfigureArgs = append(teleportNodeConfigureArgs,
+			fmt.Sprintf(`--azure-client-id=%s`, shsprintf.EscapeDefaultContext(asi.AzureClientID)))
+	}
+
+	asi.Logger.InfoContext(ctx, "Writing teleport configuration", "teleport", asi.binariesLocation.teleport, "args", teleportNodeConfigureArgs)
+	teleportNodeConfigureCmd := exec.CommandContext(ctx, asi.binariesLocation.teleport, teleportNodeConfigureArgs...)
+	teleportNodeConfigureCmdOutput, err := teleportNodeConfigureCmd.CombinedOutput()
+	if err != nil {
+		return trace.Wrap(err, string(teleportNodeConfigureCmdOutput))
+	}
+
+	asi.Logger.InfoContext(ctx, "Enabling and starting teleport service")
+	systemctlEnableNowCMD := exec.CommandContext(ctx, asi.binariesLocation.systemctl, "enable", "--now", "teleport")
+	systemctlEnableNowCMDOutput, err := systemctlEnableNowCMD.CombinedOutput()
+	if err != nil {
+		return trace.Wrap(err, string(systemctlEnableNowCMDOutput))
+	}
+
+	return nil
+}
+
+func (asi *AutodiscoverNodeInstaller) getIMDSClient(ctx context.Context) (imds.Client, error) {
+	// detect and fetch cloud provider metadata
+	imdsClient, err := cloud.DiscoverInstanceMetadata(ctx, asi.imdsProviders)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.BadParameter("Auto Discover only runs on Cloud instances with IMDS/Metadata service enabled. Ensure the service is running and try again.")
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	return imdsClient, nil
+}
+
+func (asi *AutodiscoverNodeInstaller) fetchTargetVersion(ctx context.Context) string {
```

Comment:
We have the logic for dealing with automatic updates versions in lib/automaticupgrades/channel.go. Is this something that we can reuse (or possibly extend) instead of reimplementing it here?

Reply:
We already use the `github.com/gravitational/teleport/lib/automaticupgrades/version` package.

Aside from parsing the URL from string into `net.URL` we would call this method:
```
	targetVersion, err := version.NewBasicHTTPVersionGetter(upgradeURL).GetVersion(ctx)
```

@hugoShaka Is there something else we can add to the package?